bindinfo/cache.go

```diff
@@ -99,37 +115,28 @@ func (h *Handle) Get() cache {
 
 // LoadDiff is used to load new bind info to cache bc.
 func (bindCacheUpdater *BindCacheUpdater) loadDiff(sql string, bc cache) error {
-	recordSets, err := bindCacheUpdater.ctx.(sqlexec.SQLExecutor).Execute(context.Background(), sql)
+	rows, _, err := bindCacheUpdater.ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(nil, sql)
 	if err != nil {
-		return errors.Trace(err)
+		return err
 	}
-
-	rs := recordSets[0]
-	defer terror.Call(rs.Close)
-	chkBatch := rs.NewRecordBatch()
-	for {
-		err = rs.Next(context.TODO(), chkBatch)
-		if err != nil || chkBatch.NumRows() == 0 {
-			return errors.Trace(err)
+	for _, row := range rows {
+		record := newBindRecord(row)
+		err = bc.appendNode(record, bindCacheUpdater.parser)
+		if err != nil {
+			return err
 		}
-
-		it := chunk.NewIterator4Chunk(chkBatch.Chunk)
-		for row := it.Begin(); row != it.End(); row = it.Next() {
-			record := newBindMeta(row)
-			err = bc.appendNode(record, bindCacheUpdater.parser)
-			if err != nil {
-				return err
-			}
-			if record.UpdateTime.Compare(bindCacheUpdater.lastUpdateTime) == 1 {
-				bindCacheUpdater.lastUpdateTime = record.UpdateTime
-			}
+		if record.UpdateTime.Compare(bindCacheUpdater.lastUpdateTime) == 1 {
+			bindCacheUpdater.lastUpdateTime = record.UpdateTime
 		}
 	}
+	return nil
 }
 
 // Update updates the BindCacheUpdater's cache.
 // The `fullLoad` is true only when tidb first startup, otherwise it is false.
 func (bindCacheUpdater *BindCacheUpdater) Update(fullLoad bool) (err error) {
+	bindCacheUpdater.lock.Lock()
```

Comment:
Is it needed? All the operations here are thread safe.

Reply:
after adding global bind, the bindCacheUpdater also need write the buffer, so we need a lock

Reply:
The buffer updates are atomic.